xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+        lon = float(lon)
+    elif orig_type == int:
+        lon = int(lon)
+
+    return lon
+
+
+def _get_axis_weights(xvariable,
+                      lat_bounds=None,
+                      lon_bounds=None,
+                      weight='weighted'):
+    """Function to generate axis weights for subsequent operations (e.g.,
+    area averaging) with the option to generate weights based on a rectilinear
+    regional selection.
+
+    Operations include:
+
+    - Generate weights for axes using axis bounds
+    - If a region is specified, calculate the weights corresponding to the
+    regional bounds
+
+    Parameters
+    ----------
+    xvariable : xr.DataArray
+        xr.DataArray produced using ``open_variable``
+    lat_bounds : list, optional
+        list of latitude boundaries for regional selection (e.g.,
+        lat_bounds=[-5, 5])
+    lon_bounds : list, optional
+        list of longitude boundaries for regional selection (e.g.,
+        lon_bounds=[-170, -120])
+    weight : str, optional
+        Axis weighting option (either "weighted" or "unweighted"), default
+        "weighted"
+
+    Returns
+    -------
+    dict
+        Set of weights for each eligible dimension.
+
+    Notes
+    -----
+    This is a prototype and needs work to make it work more generically. It
+    currently only works with lat / lon. It might be useful to check the
+    latitude axis units.
+
+    Examples
+    --------
+    Import and call module:
+    >>> from xcdat.dataset import open_dataset
+    >>> from xcdat.variable import open_variable
+    >>>
+    >>> filename = 'test.nc'
+    >>> ds = open_dataset(filename)
+    >>> tas = open_variable(ds, 'tas')
+    >>> axis_weights = _get_axis_weights(tas)
+
+    Raises
+    ------
+    ValueError
+        Incorrect weight option specified
+    """
+
+    # in the case bounds are unweighted, just return ones
+    if weight == 'unweighted':
+        axis_weights = {}
+        for dim in xvariable.dims:
+            # uniform weights based on original dimension
+            xdim = xvariable[dim]
+            dim_wts = xr.DataArray(np.ones(len(xdim)),
+                                   dims=xdim.dims,
+                                   coords=xdim.coords)
+            dim_wts.name = dim + '_wts'
+            axis_weights[dim] = dim_wts
+        return axis_weights
+
+    # just check a valid weight option is provided
+    # if invalid, throw error, else continue (with weighted logic)
+    # REVIEW: Should we throw a different exception type?
+    if weight not in ['weighted', 'unweighted']:
+        raise ValueError('Incorrect weight option. Choose either "weighted"' +
+                         ' or "unweighted"')
+
+    # note that xvariables.bounds does not return a list
+    # so for now this is hard-coded
+    bounds = {}
+    bounds['time'] = xvariable.bounds.time.copy(deep=True)
+    bounds['lat'] = xvariable.bounds.lat.copy(deep=True)
+    bounds['lon'] = xvariable.bounds.lon.copy(deep=True)
```

Comment:
With my recent merge of the `feature/80-open-dataset` branch, bounds are now referenced as coordinates.

I should probably turn the chunk of code below for retrieving bounds dynamically into a standalone method (cf_xarray has only one for Datasets since DataArrays weren't designed with coordinate bounds in mind).

https://github.com/tomvothecoder/xcdat/blob/1471ac8b4881379de37f9ea964596df708a745b8/xcdat/average_tv.py#L42-L54